llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp

```diff
@@ -1077,7 +1079,16 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .clampNumElements(0, v8s8, v16s8)
       .clampNumElements(0, v4s16, v8s16)
       .clampNumElements(0, v4s32, v4s32)
-      .clampNumElements(0, v2s64, v2s64);
+      .clampNumElements(0, v2s64, v2s64)
+      .bitcastIf(
```

Comment:
Instead of doing bitcast, why not make v2p0 legal? Generating bitcast and going through PTRTOINT seems sub-optimal.

Reply:
Three are two options we have, both of which should produce the same output assembly. Either we try and keep v2p0 shuffles as legal, which means having to handle them in every shuffle combine that we add and legalized all the various vNp0 types correctly, which are easy to miss. The alternative like this patch does it to convert them to v2s64 and just deal with that single type.

Because of the relative rarity of pointer shuffles it feels fine to convert them to s64 and let that legalizing happen as a s64 operation.

Reply:
Alright, thanks for the explanation.